---
title: "How to use driveR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{How to use driveR?}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

Cancer genomes contain large numbers of somatic alterations but few genes drive tumor development. Identifying molecular cancer driver genes is critical for precision oncology. Most of current approaches either identify driver genes based on mutational recurrence (across a cohort) or using estimated scores predicting the functional consequences of mutations. 

`driveR` is a tool for scoring and ranking genes from individual somatic whole exome/genome sequencing data according to cancer driverness. As features, driveR uses coding impact metaprediction scores, non-coding impact scores, somatic copy number alteration scores, hotspot gene/double-hit gene condition, 'phenolyzer' gene scores and memberships to cancer-related KEGG pathways. It uses these features to calculate cancer-type-specific driverness scores for each gene using the related task of a multi-task learning model.

Below, a step-by-step work flow for `driveR` is provided. Note that some steps require operations outside of R. The example data provided within the package are for a lung adenocarcinoma patient studied in Imielinski et al. [^1]

[^1]: Imielinski M, Greulich H, Kaplan B, et al. Oncogenic and sorafenib-sensitive ARAF mutations in lung adenocarcinoma. J Clin Invest. 2014;124(4):1582-6.

# Load the package and prepare input data

```{r setup}
library(driveR)
```

As input, `create_features_df()`, the function used to create the features table for driverness prediction, requires the following:

- `annovar_csv_path`: the `/path/to/ANNOVAR/csv/output/file`
- `scna_df`: a data frame containing SCNA segments, containing the columns "chr", "start", "end" and "log2ratio"
- `phenolyzer_annotated_gene_list_path`: the `/path/to/phenolyzer/annotated_gene_list/output/file`

## ANNOVAR CSV output
We first run ANNOVAR. An example command provided below:

```{bash, eval=FALSE}
table_annovar.pl example.avinput ~/annovar/humandb/ -buildver hg19 -out /path/to/ouput -remove -protocol refGene,cytoBand,exac03,avsnp147,dbnsfp30a,cosmic91_coding,cosmic91_noncoding -operation gx,r,f,f,f,f,f -nastring . -csvout -polish
```

The required filters are, as listed in the above command, `refGene,cytoBand,exac03,avsnp147,dbnsfp30a,cosmic91_coding,cosmic91_noncoding`.

With the package, an example (modified) ANNOVAR csv output file is available:

```{r example_av_csv}
path2annovar_csv <- system.file("extdata/example.hg19_multianno.csv",
                                 package = "driveR")
```

## SCNA table

Next, we prepare a SCNA data frame (GRCh37 is required). Again, an example data frame is provided within the package:

```{r example_scna}
head(example_scna_table)
```

## phenolyzer output

Finally, for the phenolyzer annotated_gene_list output file, we first obtain the genes to be scored using `create_features_df` and setting `prep_phenolyzer_input=TRUE`:

```{r phenolyzer_input}
phenolyzer_genes <- create_features_df(annovar_csv_path = path2annovar_csv,
                                       scna_df = example_scna_table,
                                       prep_phenolyzer_input = TRUE)
```

Next, we save these genes to be used as input for phenolyzer:
```{r save_phenolyzer_input, eval=FALSE}
write.table(x = data.frame(gene = phenolyzer_genes), 
            file = "input_genes.txt", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)
```

We create another text file, named `phenolyzer_disease.txt`, containing the phenotype (i.e. cancer type. in this case "lung adenocarcinoma") to be used for scoring with phenolyzer. An example command for running phenolyzer is provided below:

```{bash, eval=FALSE}
perl ~/phenolyzer/disease_annotation.pl -f phenolyzer_disease.txt -prediction -phenotype -logistic --gene input_genes.txt -out phenolyzer_out/example
```

This should produce, among other outputs, the annotated_gene_list output file. An example annotated_gene_list output file is provided within the package:
```{r example_phenoylzer}
path2phenolyzer_out <- system.file("extdata/example.annotated_gene_list",
                                    package = "driveR")
```

# Create the features data frame

After creating the necessary input data (as detailed above), we simply run `create_features_df()` to obtain the features data frame:

```{r features_df}
features_df <- create_features_df(annovar_csv_path = path2annovar_csv,
                                  scna_df = example_scna_table, 
                                  phenolyzer_annotated_gene_list_path = path2phenolyzer_out)
```

# Calculate cancer driverness probabilities

Finally, we can calculate the cancer driverness probability of each gene using `calculate_driverness_probs()`. For this function, `cancer_type` can be of the short names for the 21 different cancer types that was used to train the multi-task learning model utilized in this package:

```{r cancer_types}
knitr::kable(MTL_submodel_descriptions)
```

Below is the example run for the lung adenocarcinoma patient:

```{r driver_prob}
driver_prob_df <- calculate_driverness_probs(features_df = features_df,
                                             cancer_type = "LUAD")
head(driver_prob_df, 10)
```

The function returns a data frame of genes with cancer driverness probabilities (using the selected sub-model of a multi-task learning model trained using 21 different cancer types). In the above example, the top 10 genes contain recognizable cancer driver genes.
