<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create Data Frame of Features for Driver Gene Prioritization — create_features_df • driveR</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="../docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Create Data Frame of Features for Driver Gene Prioritization — create_features_df"><meta property="og:description" content="Create Data Frame of Features for Driver Gene Prioritization"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">driveR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.0.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/how_to_use.html">How to use driveR</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/egeulgen/driveR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create Data Frame of Features for Driver Gene Prioritization</h1>
    <small class="dont-index">Source: <a href="https://github.com/egeulgen/driveR/blob/HEAD/R/core_functions.R" class="external-link"><code>R/core_functions.R</code></a></small>
    <div class="hidden name"><code>create_features_df.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Create Data Frame of Features for Driver Gene Prioritization</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_features_df</span><span class="op">(</span></span>
<span>  <span class="va">annovar_csv_path</span>,</span>
<span>  <span class="va">scna_df</span>,</span>
<span>  <span class="va">phenolyzer_annotated_gene_list_path</span>,</span>
<span>  batch_analysis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prep_phenolyzer_input <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  build <span class="op">=</span> <span class="st">"GRCh37"</span>,</span>
<span>  log2_ratio_threshold <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  gene_overlap_threshold <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  MCR_overlap_threshold <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  hotspot_threshold <span class="op">=</span> <span class="fl">5L</span>,</span>
<span>  log2_hom_loss_threshold <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  na.string <span class="op">=</span> <span class="st">"."</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>annovar_csv_path</dt>
<dd><p>path to 'ANNOVAR' csv output file</p></dd>


<dt>scna_df</dt>
<dd><p>the SCNA segments data frame. Must contain:</p><dl><dt>chr</dt>
<dd><p>chromosome the segment is located in</p></dd>

<dt>start</dt>
<dd><p>start position of the segment</p></dd>

<dt>end</dt>
<dd><p>end position of the segment</p></dd>

<dt>log2ratio</dt>
<dd><p>log<sub>2</sub> ratio of the segment</p></dd>


</dl></dd>


<dt>phenolyzer_annotated_gene_list_path</dt>
<dd><p>path to 'phenolyzer'
"annotated_gene_list" file</p></dd>


<dt>batch_analysis</dt>
<dd><p>boolean to indicate whether to perform batch analysis
(<code>TRUE</code>, default) or personalized analysis (<code>FALSE</code>). If <code>TRUE</code>,
a column named 'tumor_id' should be present in both the ANNOVAR csv and the SCNA
table.</p></dd>


<dt>prep_phenolyzer_input</dt>
<dd><p>boolean to indicate whether or not to create
a vector of genes for use as the input of 'phenolyzer' (default = <code>FALSE</code>).
If <code>TRUE</code>, the features data frame is not created and instead the vector
of gene symbols (union of all genes for which scores are available) is
returned.</p></dd>


<dt>build</dt>
<dd><p>genome build for the SCNA segments data frame (default = "GRCh37")</p></dd>


<dt>log2_ratio_threshold</dt>
<dd><p>the log<sub>2</sub>
ratio threshold for keeping high-confidence SCNA events (default = 0.25)</p></dd>


<dt>gene_overlap_threshold</dt>
<dd><p>the percentage threshold for the overlap between
a segment and a transcript (default = 25). This means that if only a segment
overlaps a transcript more than this threshold, the transcript is assigned
the segment's SCNA event.</p></dd>


<dt>MCR_overlap_threshold</dt>
<dd><p>the percentage threshold for the overlap between
a gene and an MCR region (default = 25). This means that if only a gene
overlaps an MCR region more than this threshold, the gene is assigned the
SCNA density of the MCR</p></dd>


<dt>hotspot_threshold</dt>
<dd><p>to determine hotspot genes, the (integer) threshold
for the minimum number of cases with certain mutation in COSMIC (default = 5)</p></dd>


<dt>log2_hom_loss_threshold</dt>
<dd><p>to determine double-hit events, the
log<sub>2</sub> threshold for identifying
homozygous loss events (default = -1).</p></dd>


<dt>verbose</dt>
<dd><p>boolean controlling verbosity (default = <code>TRUE</code>)</p></dd>


<dt>na.string</dt>
<dd><p>string that was used to indicate when a score is not available
during annotation with ANNOVAR (default = ".")</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>If <code>prep_phenolyzer_input=FALSE</code> (default), a data frame of
features for prioritizing cancer driver genes (<code>gene_symbol</code> as
the first column and 26 other columns containing features). If
<code>prep_phenolyzer_input=TRUE</code>, the functions returns a vector gene symbols
(union of all gene symbols for which scores are available) to be used as the
input for performing 'phenolyzer' analysis.</p>


<p>The features data frame contains the following columns:</p><dl><dt>gene_symbol</dt>
<dd><p>HGNC gene symbol</p></dd>

<dt>metaprediction_score</dt>
<dd><p>the maximum metapredictor (coding) impact score for the gene</p></dd>

<dt>noncoding_score</dt>
<dd><p>the maximum non-coding PHRED-scaled CADD score for the gene</p></dd>

<dt>scna_score</dt>
<dd><p>SCNA proxy score. SCNA density (SCNA/Mb) of the minimal common region (MCR) in which the gene is located</p></dd>

<dt>hotspot_double_hit</dt>
<dd><p>boolean indicating whether the gene is a hotspot gene (indication of oncogenes) or subject to double-hit (indication of tumor-suppressor genes)</p></dd>

<dt>phenolyzer_score</dt>
<dd><p>'phenolyzer' score for the gene</p></dd>

<dt>hsa03320</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04010</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04020</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04024</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04060</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04066</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04110</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04115</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04150</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04151</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04210</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04310</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04330</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04340</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04350</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04370</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04510</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04512</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04520</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04630</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>

<dt>hsa04915</dt>
<dd><p>boolean indicating whether or not the gene takes part in this KEGG pathway</p></dd>


</dl></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="prioritize_driver_genes.html">prioritize_driver_genes</a></code> for prioritizing cancer driver genes</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="va">path2annovar_csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example.hg19_multianno.csv"</span>,</span></span>
<span class="r-in"><span>                                package <span class="op">=</span> <span class="st">"driveR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">path2phenolyzer_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/example.annotated_gene_list"</span>,</span></span>
<span class="r-in"><span>                                   package <span class="op">=</span> <span class="st">"driveR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">features_df</span> <span class="op">&lt;-</span> <span class="fu">create_features_df</span><span class="op">(</span>annovar_csv_path <span class="op">=</span> <span class="va">path2annovar_csv</span>,</span></span>
<span class="r-in"><span>                                  scna_df <span class="op">=</span> <span class="va">example_scna_table</span>,</span></span>
<span class="r-in"><span>                                  phenolyzer_annotated_gene_list_path <span class="op">=</span> <span class="va">path2phenolyzer_out</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predicting impact of coding variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Predicting impact of non-coding variants</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Determining gene-level SCNAs (This may take a while)</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>replacing previous import ‘S4Arrays::makeNindexFromArrayViewport’ by ‘DelayedArray::makeNindexFromArrayViewport’ when loading ‘SummarizedExperiment’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Scoring SCNA events</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Determining hotspot/double-hit genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Parsing 'phenolyzer' gene scores</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Assessing memberships to KEGG - cancer-related pathways</span>
<span class="r-in"><span><span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ege Ulgen.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '21b3b36a9d378c47ddeef07e523fb44b',
    indexName: 'egeulgen_driver',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

