<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>How to use driveR • driveR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="How to use driveR">
<meta property="og:description" content="driveR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">driveR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/how_to_use.html">How to use driveR</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/egeulgen/driveR/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="how_to_use_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>How to use driveR</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/egeulgen/driveR/blob/master/vignettes/how_to_use.Rmd"><code>vignettes/how_to_use.Rmd</code></a></small>
      <div class="hidden name"><code>how_to_use.Rmd</code></div>

    </div>

    
    
<p>Cancer genomes contain large numbers of somatic alterations but few genes drive tumor development. Identifying molecular cancer driver genes is critical for precision oncology. Most of current approaches either identify driver genes based on mutational recurrence (across a cohort) or using estimated scores predicting the functional consequences of mutations.</p>
<p><code>driveR</code> is a tool for scoring and ranking genes from individual somatic whole exome/genome sequencing data according to cancer driverness. As features, driveR uses coding impact metaprediction scores, non-coding impact scores, somatic copy number alteration scores, hotspot gene/double-hit gene condition, ‘phenolyzer’ gene scores and memberships to cancer-related KEGG pathways. It uses these features to calculate cancer-type-specific driverness scores for each gene using the related task of a multi-task learning model.</p>
<p>Below, a step-by-step work flow for <code>driveR</code> is provided. Note that some steps require operations outside of R. The example data provided within the package are for a lung adenocarcinoma patient studied in Imielinski et al. <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<div id="load-the-package-and-prepare-input-data" class="section level1">
<h1 class="hasAnchor">
<a href="#load-the-package-and-prepare-input-data" class="anchor"></a>Load the package and prepare input data</h1>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">driveR</span>)</pre></body></html></div>
<p>As input, <code><a href="../reference/create_features_df.html">create_features_df()</a></code>, the function used to create the features table for driverness prediction, requires the following:</p>
<ul>
<li>
<code>annovar_csv_path</code>: the <code>/path/to/ANNOVAR/csv/output/file</code>
</li>
<li>
<code>scna_df</code>: a data frame containing SCNA segments, containing the columns “chr”, “start”, “end” and “log2ratio”</li>
<li>
<code>phenolyzer_annotated_gene_list_path</code>: the <code>/path/to/phenolyzer/annotated_gene_list/output/file</code>
</li>
</ul>
<div id="annovar-csv-output" class="section level2">
<h2 class="hasAnchor">
<a href="#annovar-csv-output" class="anchor"></a>ANNOVAR CSV output</h2>
<p>We first run ANNOVAR. An example command provided below:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="ex">table_annovar.pl</span> example.avinput ~/annovar/humandb/ -buildver hg19 -out /path/to/ouput -remove -protocol refGene,cytoBand,exac03,avsnp147,dbnsfp30a,cosmic91_coding,cosmic91_noncoding -operation gx,r,f,f,f,f,f -nastring . -csvout -polish</span></code></pre></div>
<p>The required filters are, as listed in the above command, <code>refGene,cytoBand,exac03,avsnp147,dbnsfp30a,cosmic91_coding,cosmic91_noncoding</code>.</p>
<p>With the package, an example (modified) ANNOVAR csv output file is available:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">path2annovar_csv</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/example.hg19_multianno.csv"</span>,
                                 <span class="kw">package</span> <span class="kw">=</span> <span class="st">"driveR"</span>)</pre></body></html></div>
</div>
<div id="scna-table" class="section level2">
<h2 class="hasAnchor">
<a href="#scna-table" class="anchor"></a>SCNA table</h2>
<p>Next, we prepare a SCNA data frame (GRCh37 is required). Again, an example data frame is provided within the package:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">example_scna_table</span>)
<span class="co">#&gt;   chr   start     end log2ratio</span>
<span class="co">#&gt; 1   1       1    9001 0.8875253</span>
<span class="co">#&gt; 2   1    9002   10001 0.6959938</span>
<span class="co">#&gt; 3   1   10002 3695599 0.3785116</span>
<span class="co">#&gt; 4   1 3695600 4715896 0.4646683</span>
<span class="co">#&gt; 5   1 4715897 4717788 0.4005379</span>
<span class="co">#&gt; 6   1 4717789 5798840 0.4854268</span></pre></body></html></div>
</div>
<div id="phenolyzer-output" class="section level2">
<h2 class="hasAnchor">
<a href="#phenolyzer-output" class="anchor"></a>phenolyzer output</h2>
<p>Finally, for the phenolyzer annotated_gene_list output file, we first obtain the genes to be scored using <code>create_features_df</code> and setting <code>prep_phenolyzer_input=TRUE</code>:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">phenolyzer_genes</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/create_features_df.html">create_features_df</a></span>(<span class="kw">annovar_csv_path</span> <span class="kw">=</span> <span class="no">path2annovar_csv</span>,
                                       <span class="kw">scna_df</span> <span class="kw">=</span> <span class="no">example_scna_table</span>,
                                       <span class="kw">prep_phenolyzer_input</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">#&gt; </span></pre></body></html></div>
<p>Next, we save these genes to be used as input for phenolyzer:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">gene</span> <span class="kw">=</span> <span class="no">phenolyzer_genes</span>),
            <span class="kw">file</span> <span class="kw">=</span> <span class="st">"input_genes.txt"</span>,
            <span class="kw">row.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">col.names</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">quote</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<p>We create another text file, named <code>phenolyzer_disease.txt</code>, containing the phenotype (i.e. cancer type. in this case “lung adenocarcinoma”) to be used for scoring with phenolyzer. An example command for running phenolyzer is provided below:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">perl</span> ~/phenolyzer/disease_annotation.pl -f phenolyzer_disease.txt -prediction -phenotype -logistic --gene input_genes.txt -out phenolyzer_out/example</span></code></pre></div>
<p>This should produce, among other outputs, the annotated_gene_list output file. An example annotated_gene_list output file is provided within the package:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">path2phenolyzer_out</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata/example.annotated_gene_list"</span>,
                                    <span class="kw">package</span> <span class="kw">=</span> <span class="st">"driveR"</span>)</pre></body></html></div>
</div>
</div>
<div id="create-the-features-data-frame" class="section level1">
<h1 class="hasAnchor">
<a href="#create-the-features-data-frame" class="anchor"></a>Create the features data frame</h1>
<p>After creating the necessary input data (as detailed above), we simply run <code><a href="../reference/create_features_df.html">create_features_df()</a></code> to obtain the features data frame:</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">features_df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/create_features_df.html">create_features_df</a></span>(<span class="kw">annovar_csv_path</span> <span class="kw">=</span> <span class="no">path2annovar_csv</span>,
                                  <span class="kw">scna_df</span> <span class="kw">=</span> <span class="no">example_scna_table</span>,
                                  <span class="kw">phenolyzer_annotated_gene_list_path</span> <span class="kw">=</span> <span class="no">path2phenolyzer_out</span>)</pre></body></html></div>
</div>
<div id="calculate-cancer-driverness-probabilities" class="section level1">
<h1 class="hasAnchor">
<a href="#calculate-cancer-driverness-probabilities" class="anchor"></a>Calculate cancer driverness probabilities</h1>
<p>Finally, we can calculate the cancer driverness probability of each gene using <code><a href="../reference/calculate_driverness_probs.html">calculate_driverness_probs()</a></code>. For this function, <code>cancer_type</code> can be of the short names for the 21 different cancer types that was used to train the multi-task learning model utilized in this package:</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="no">MTL_submodel_descriptions</span>)</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">short_name</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">BLCA</td>
<td align="left">Bladder Urothelial Cancer</td>
</tr>
<tr class="even">
<td align="left">BRCA</td>
<td align="left">Breast Cancer</td>
</tr>
<tr class="odd">
<td align="left">CESC</td>
<td align="left">Cervical Squamous Cell Carcinoma</td>
</tr>
<tr class="even">
<td align="left">COAD</td>
<td align="left">Colon Adenocarcinoma</td>
</tr>
<tr class="odd">
<td align="left">GBM</td>
<td align="left">Brain Glioblastoma Multiforme</td>
</tr>
<tr class="even">
<td align="left">HNSC</td>
<td align="left">Head and Neck Squamous Cell Carcinoma</td>
</tr>
<tr class="odd">
<td align="left">KIRC</td>
<td align="left">Kidney Renal Clear Cell Carcinoma</td>
</tr>
<tr class="even">
<td align="left">KIRP</td>
<td align="left">Kidney Renal Papillary Cell Carcinoma</td>
</tr>
<tr class="odd">
<td align="left">LAML</td>
<td align="left">Acute Myeloid Leukemia</td>
</tr>
<tr class="even">
<td align="left">LGG</td>
<td align="left">Brain Lower Grade Glioma</td>
</tr>
<tr class="odd">
<td align="left">LIHC</td>
<td align="left">Liver Hepatocellular carcinoma</td>
</tr>
<tr class="even">
<td align="left">LUAD</td>
<td align="left">Lung Adenocarcinoma</td>
</tr>
<tr class="odd">
<td align="left">LUSC</td>
<td align="left">Lung Squamous Cell Carcinoma</td>
</tr>
<tr class="even">
<td align="left">OV</td>
<td align="left">Ovarian Serous Cystadenocarcinoma</td>
</tr>
<tr class="odd">
<td align="left">PAAD</td>
<td align="left">Pancreatic Cancer</td>
</tr>
<tr class="even">
<td align="left">PRAD</td>
<td align="left">Prostate Adenocarcinoma</td>
</tr>
<tr class="odd">
<td align="left">READ</td>
<td align="left">Rectum Adenocarcinoma</td>
</tr>
<tr class="even">
<td align="left">SKCM</td>
<td align="left">Skin Cutaneous melanoma</td>
</tr>
<tr class="odd">
<td align="left">STAD</td>
<td align="left">Gastric Adenocarcinoma</td>
</tr>
<tr class="even">
<td align="left">THCA</td>
<td align="left">Head and Neck Thyroid Carcinoma</td>
</tr>
<tr class="odd">
<td align="left">UCEC</td>
<td align="left">Uterine Corpus Endometrial Carcinoma</td>
</tr>
</tbody>
</table>
<p>Below is the example run for the lung adenocarcinoma patient:</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">driver_prob_df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/calculate_driverness_probs.html">calculate_driverness_probs</a></span>(<span class="kw">features_df</span> <span class="kw">=</span> <span class="no">features_df</span>,
                                             <span class="kw">cancer_type</span> <span class="kw">=</span> <span class="st">"LUAD"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">driver_prob_df</span>, <span class="fl">10</span>)
<span class="co">#&gt;      gene_symbol driverness_prob</span>
<span class="co">#&gt; 842         TP53       0.9996252</span>
<span class="co">#&gt; 1966       EP300       0.9995983</span>
<span class="co">#&gt; 3211       CCND3       0.9990539</span>
<span class="co">#&gt; 3654        EGFR       0.9972283</span>
<span class="co">#&gt; 3572      PIK3R2       0.9937462</span>
<span class="co">#&gt; 2297       PRKCA       0.9896174</span>
<span class="co">#&gt; 653        THBS1       0.9813223</span>
<span class="co">#&gt; 4342        SOS2       0.9803484</span>
<span class="co">#&gt; 1030         EGF       0.9717150</span>
<span class="co">#&gt; 4589       ERBB2       0.9688511</span></pre></body></html></div>
<p>The function returns a data frame of genes with cancer driverness probabilities (using the selected sub-model of a multi-task learning model trained using 21 different cancer types). In the above example, the top 10 genes contain recognizable cancer driver genes.</p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Imielinski M, Greulich H, Kaplan B, et al. Oncogenic and sorafenib-sensitive ARAF mutations in lung adenocarcinoma. J Clin Invest. 2014;124(4):1582-6.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Ege Ulgen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
